File: admin-backend/admin-shared/admin-shared-persistence/src/main/java/io/github/faustofan/admin/shared/persistence/interceptor/AuditDraftInterceptor.java
package io.github.faustofan.admin.shared.persistence.interceptor;

import io.github.faustofan.admin.shared.common.context.AppContextHolder;
import io.github.faustofan.admin.shared.common.exception.CommonErrorCode;
import io.github.faustofan.admin.shared.common.exception.SystemException;
import io.github.faustofan.admin.shared.persistence.model.BaseEntity;
import io.github.faustofan.admin.shared.persistence.model.BaseEntityDraft;
import io.github.faustofan.admin.shared.persistence.model.BaseEntityProps;
import org.babyfish.jimmer.ImmutableObjects;
import org.babyfish.jimmer.sql.DraftInterceptor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.springframework.stereotype.Component;

import java.time.Instant;

/**
 * 审计字段拦截器
 * 在保存草稿前，自动设置创建人、创建时间、更新人、更新时间等审计字段
 */
@Component
public class AuditDraftInterceptor implements DraftInterceptor<BaseEntity, BaseEntityDraft> {

    /**
     * 在保存草稿前调用
     * @param draft 草稿对象
     * @param original 原始实体对象，新增时为 null
     */
    @Override
    public void beforeSave(@NotNull BaseEntityDraft draft, @Nullable BaseEntity original) {
        if(null == AppContextHolder.getContext())
            throw new SystemException(CommonErrorCode.CONTEXT_MISSING);

        Instant now = Instant.now();

        // 新增时，设置创建人和创建时间
        if (original == null) {
            if(!ImmutableObjects.isLoaded(draft, BaseEntityProps.CREATED_TIME)) {
                draft.setCreatedTime(now);
            }
            if(!ImmutableObjects.isLoaded(draft, BaseEntityProps.DELETED)) {
                draft.setDeleted(false); // 默认未删除
            }
        }

        draft.setUpdatedTime(now);
    }
}

File: src/main/kotlin/io/github/faustofan/admin/config/jimmer/AuditDraftInterceptor.kt
package io.github.faustofan.admin.config.jimmer

import io.github.faustofan.admin.common.context.AppContextHolder
import io.github.faustofan.admin.common.model.BaseEntity
import io.github.faustofan.admin.common.model.BaseEntityDraft
import org.babyfish.jimmer.kt.isLoaded
import org.babyfish.jimmer.sql.DraftInterceptor
import org.springframework.stereotype.Component
import java.time.Instant

/**
 * 审计字段拦截器
 *
 * 在实体保存前自动填充审计字段：
 * - createdTime: 创建时间
 * - updatedTime: 更新时间
 * - createdBy: 创建人ID
 * - updatedBy: 更新人ID
 */
@Component
class AuditDraftInterceptor : DraftInterceptor<BaseEntity, BaseEntityDraft> {

    override fun beforeSave(draft: BaseEntityDraft, original: BaseEntity?) {
        val now = Instant.now()
        val currentUserId = AppContextHolder.get().userId

        // 新增时填充创建时间和创建人
        if (original === null) {
            if (!isLoaded(draft, BaseEntity::createdTime)) {
                draft.createdTime = now
            }
            if (!isLoaded(draft, BaseEntity::createdBy)) {
                draft.createdBy = currentUserId
            }
            // 新增时逻辑删除字段默认为 false
            if (!isLoaded(draft, BaseEntity::deleted)) {
                draft.deleted = false
            }
        }

        // 新增和更新都填充更新时间和更新人
        draft.updatedTime = now
        draft.updatedBy = currentUserId
    }
}

